### ISSUE TITLE
How to quantize Mobilenet v2 ? 

### ISSUE THREAD

**Issue by WenguoLi** [tensorflow/tensorflow#19014]
## System information

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):no
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):Ubuntu 16.04
- TensorFlow installed from (source or binary):Source
- TensorFlow version (use command below):1.8.0rc0
- Python version: 2.7.12
- Bazel version (if compiling from source): 0.12.0
- GCC/Compiler version (if compiling from source): 5.4.0
- CUDA/cuDNN version:cuda-9.0/7.0
- GPU model and memory:GeForce GTX 1080/8105MiB
- Phone: xiaomi5 (Snapdragon 820)

## Describe the problem
Using **tf.contrib.quantize.create_training_graph()** and **tf.contrib.quantize.create_eval_graph()**,  I 
quantized training  Mobilenet v2,  and  export inference graph,  but  when toco pb to tflite,  I encountered the following error:

2018-05-02 10:26:27.215975: F tensorflow/contrib/lite/toco/tooling_util.cc:1464] Array MobilenetV2/expanded_conv_2/add, which is an input to the Conv operator producing the output array MobilenetV2/expanded_conv_3/expand/Relu6, is lacking min/max data, which is necessary for quantization. Either target a non-quantized output format, or change the input graph to contain min/max information, or pass --default_ranges_min= and --default_ranges_max= if you do not care about the accuracy of results.

This layer MobilenetV2/expanded_conv_2/add  cannot be quantized ? 

Who can explain why ? 

## Source code / logs

bazel run --config=opt \
>   //tensorflow/contrib/lite/toco:toco -- \
>   --input_file=${TRAIN_DIR}/frozen_graph.pb \
>   --output_file=${TRAIN_DIR}/tflite_model.tflite \
>   --input_format=TENSORFLOW_GRAPHDEF \
>   --output_format=TFLITE \
>   --inference_type=QUANTIZED_UINT8 \
>   --input_shape=1,224,224,3 \
>   --input_array=input \
>   --output_array=MobilenetV2/Predictions/Reshape_1 \
>   --std_value=127 \
>   --mean_value=128 \
>   --dump_graphviz=${TRAIN_DIR}

```
INFO: Running command line: bazel-bin/tensorflow/contrib/lite/toco/toco '--input_file=/home/apuser/tmp/models/flowers/mobilenet_v2_1.0_224_quan/frozen_graph.pb' '--output_file=/home/apuser/tmp/models/flowers/mobilenet_v2_1.0_224_quan/tflite_model.tflite' '--input_format=TENSORFLOW_GRAPHDEF' '--output_format=TFLITE' '--inference_type=QUANTIZED_UINT8' '--input_shape=1,224,224,3' '--input_array=input' '--output_array=MobilenetV2/Predictions/Reshape_1' '--std_value=127' '--mean_value=128' '--dump_graphviz=/home/apuser/tmp/models/flowers/mobilenet_v2_1.0_224_quan'
2018-05-02 10:26:26.887728: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 1207 operators, 1752 arrays (0 quantized)
2018-05-02 10:26:26.934465: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 1207 operators, 1752 arrays (0 quantized)
2018-05-02 10:26:27.211718: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 120 operators, 228 arrays (1 quantized)
2018-05-02 10:26:27.214013: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 120 operators, 228 arrays (1 quantized)
2018-05-02 10:26:27.215023: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After pre-quantization graph transformations pass 1: 67 operators, 175 arrays (1 quantized)
2018-05-02 10:26:27.215975: F tensorflow/contrib/lite/toco/tooling_util.cc:1464] Array MobilenetV2/expanded_conv_2/add, which is an input to the Conv operator producing the output array MobilenetV2/expanded_conv_3/expand/Relu6, is lacking min/max data, which is necessary for quantization. Either target a non-quantized output format, or change the input graph to contain min/max information, or pass --default_ranges_min= and --default_ranges_max= if you do not care about the accuracy of results.
```

**Comment by WenguoLi** [comment]
I used the same method, toco quantized mobilenet_v1 is no problem.
I don't know where the problem is .
Who can help ?

**Comment by suharshs** [comment]
This looks like a recent breakage. I will get a fix soon. Thanks!

**Comment by WenguoLi** [comment]
@suharshs , Thanks , I am looking forward to your reply!

**Comment by kanul** [comment]
@suharshs, Has not the issue been resolved yet?

I'm looking forward to fix the issue as soon as possible.

If the issue has not been resolved yet, How long does it take to fix the problem?

**Comment by suharshs** [comment]
Hi @kanul,

The issue originally described should be resolved.

It seems that you are using a mobilenet_v2 implementation contains an unsupported operation. What mobilenet_v2 implementation are you using? The code I have tried out is here: https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet_v2.py .

Make sure you are freezing an eval graph and not a training graph and provide the correct input and output nodes when freezing as well.

**Comment by kanul** [comment]
@suharshs, 

I want to develop segmentation mask based on mobilnetV2 for deeplabV3

My interested task is not classification but image segmentation

https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md
Checkpoint name: mobilenetv2_coco_voc_trainaug

**Comment by OdingdongO** [comment]
In the conversion of mobilenet v2, I encountered the same problem with @WenguoLi ,but the mobilenetv1 is ok, how do you solve it? @suharshs

**Comment by suharshs** [comment]
Hi @OdingdongO , this should be fixed in 01a70dc. What TF version are you using?

**Comment by WenguoLi** [comment]
@suharshs Thanks !

**Comment by OdingdongO** [comment]
i update the tf-nightly-gpu by pip,but i got a error when i use the create_eval_ graph() in mobilenetv2
,but the mobilenetv1 is  ok . @suharshs 


tensorflow.python.framework.errors_impl.NotFoundError: Key MobilenetV2/expanded_conv_11/post_activation_bypass_quant/max not found in checkpoint
	 [[Node: save/RestoreV2 = RestoreV2[dtypes=[DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, ..., DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_FLOAT, DT_INT64], _device="/job:localhost/replica:0/task:0/device:CPU:0"](_arg_save/Const_0_0, save/RestoreV2/tensor_names, save/RestoreV2/shape_and_slices)]]
	 [[Node: save/RestoreV2/_301 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device_incarnation=1, tensor_name="edge_306_save/RestoreV2", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/device:GPU:0"]()]]

**Comment by suharshs** [comment]
@OdingdongO Where are you getting the checkpoint? Is it possible to provide a repro case?

**Comment by OdingdongO** [comment]
i can quant the mobilev2 by According to @parvizp tips now,thank you all @suharshs @parvizp

**Comment by G-mel** [comment]
I also kept on getting similar errors with using a MobileNetV2 backbone for my network.

To solve my issue, I made sure to create the training quantization nodes (`tf.contrib.quantize.create_training_graph(quant_delay=quant_delay)`) in my training graph with MobileNetV2 training scope:
```
with tf.contrib.slim.arg_scope(mobilenet_v2.training_scope()):
    <create mobilenetv2 based network>
```

Where when I created my eval graph with quantization nodes (`tf.contrib.quantize.create_eval_graph()`), I made sure to load my MobileNetv2 model **without** the above training scope. This solved all of my errors of having nodes without having min/max information. 

I separated my training and evaluation into different scripts/functions in order to make it easier to separate the two graphs. 

I then [froze](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) my saved checkpoint weights from training with the graph from evaluation, then optimized them for [inference](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py), and finally converted to TFLite with [TOCO](https://github.com/tensorflow/tensorflow/blob/bfcfad55b7b3fa4a1093fa748d4241f9457b2a84/tensorflow/contrib/lite/python/tflite_convert.py) to get a quantized tflite file.

Hope this helps.

**Comment by Yayong-guan** [comment]
@G-mel , can you quantize MobileNetV2 with multi gpu trainning,  i can  quantize with only one gpu，multi gpu will get NotFoundError key.

**Comment by apple1238033** [comment]
i get this error ,when i use export inference graph.py to produce eval.pbtxt,i found whether i use tf.contrib.quantize.create_eval_graph() or not ,the produced eval.pbtxt are totally the same ,the function tf.contrib.quantize.create_eval_graph() not work at all !!

**Comment by guvcolie** [comment]
@G-mel Thank you for your experiment! I am training a network whose backbone is mobilenetV2.
When you use tf.contrib.quantize.create_training_graph(quant_delay=quant_delay), how many GPUs do you use?

**Comment by G-mel** [comment]
@guvcolie My work was with only 1 GPU

**Comment by WindQAQ** [comment]
@parvizp I am currently working on quantization and try to run your script. However, there is a huge accuracy drop (~7%) after quantization. I am wondering how much performance drop your model suffers. Thank you!

**Comment by SteveIb** [comment]
Im trying to convert pb to tflite, i run my model using Mobilenet v2, 
when trying to convert im getting the following error.

bazel-bin/tensorflow/contrib/lite/toco/toco --input_file=/home/dell/TFObject-detection/tensorflow/models/research/object_detection/Output/frozen_inference_graph.pb --input_format=TENSORFLOW_GRAPHDEF --output_format=TFLITE --output_file=frozen_inference_graph.lite --inference_type=FLOAT --input_type=FLOAT --input_arrays=ImageTensor --output_arrays=SemanticPredictions --input_shapes=1,512,512,3 --allow_nonexistent_arrays


2018-11-15 09:54:38.341689: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 0 operators, 2 arrays (0 quantized)
2018-11-15 09:54:38.341711: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:135] Model is empty!!!
2018-11-15 09:54:38.341737: W tensorflow/contrib/lite/toco/tooling_util.cc:1246] Fixing constant output array SemanticPredictions by inserting a copy. This is not optimal.
2018-11-15 09:54:38.341749: F ./tensorflow/contrib/lite/toco/model.h:1984] Check failed: has_shape() 
Aborted (core dumped)


is that related to this issue? any hint?

**Comment by dlml** [comment]
![default](https://user-images.githubusercontent.com/35597616/49368198-0be3d300-f729-11e8-89b3-b05ab3767ea1.png)
i use" tf.contrib.quantize.create_training_graph(input_graph=g,quant_delay=100)" to train model and "tf.contrib.quantize.create_eval_graph(input_graph=g)" to eval model, but still need min and max，i don‘t kown why？ @suharshs

**Comment by zuqiutxy** [comment]
@WindQAQ  hi,dou u successfully use parvizp's script,could you tell me which version of tensorflow do you ues. I use parvizp's script, it also occured the error"lacking min and max information for depthwise/Relu6"

**Comment by desimmons** [comment]
> I also kept on getting similar errors with using a MobileNetV2 backbone for my network.
> 
> To solve my issue, I made sure to create the training quantization nodes (`tf.contrib.quantize.create_training_graph(quant_delay=quant_delay)`) in my training graph with MobileNetV2 training scope:
> 
> ```
> with tf.contrib.slim.arg_scope(mobilenet_v2.training_scope()):
>     <create mobilenetv2 based network>
> ```
> Where when I created my eval graph with quantization nodes (`tf.contrib.quantize.create_eval_graph()`), I made sure to load my MobileNetv2 model **without** the above training scope. This solved all of my errors of having nodes without having min/max information.
> 
> I separated my training and evaluation into different scripts/functions in order to make it easier to separate the two graphs.
> 
> I then [froze](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) my saved checkpoint weights from training with the graph from evaluation, then optimized them for [inference](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference.py), and finally converted to TFLite with [TOCO](https://github.com/tensorflow/tensorflow/blob/bfcfad55b7b3fa4a1093fa748d4241f9457b2a84/tensorflow/contrib/lite/python/tflite_convert.py) to get a quantized tflite file.
> 
> Hope this helps.

I found this response sufficient to resolve this issue. It is important to remember, however, that restoring the eval graph should be done **after** create_eval_graph() has been called
https://github.com/tensorflow/tensorflow/issues/18919#issuecomment-386771852

**Comment by NorwayLobster** [comment]
> Who can help ?

您好，　你的问题解决了吗？？？

**Comment by bhavints** [comment]
> Hi @OdingdongO , this should be fixed in [01a70dc](https://github.com/tensorflow/tensorflow/commit/01a70dc43d32eb5add5f1cb5de2d6c98ed88dd83). What TF version are you using?

"What TF" << I read that differently than intended ;)
